File: scripts/apply-migrations.py
#!/usr/bin/env python3
"""Apply Supabase migrations via the Management API.

Reads token + project ref from .env.tools.local, POSTs each SQL file
to https://api.supabase.com/v1/projects/{ref}/database/query, and
reports outcome per file.
"""
import json
import sys
import urllib.request
import urllib.error
from pathlib import Path

def load_env(path: Path) -> dict:
    out = {}
    if not path.exists():
        return out
    for line in path.read_text().splitlines():
        line = line.strip()
        if not line or line.startswith("#") or "=" not in line:
            continue
        k, v = line.split("=", 1)
        out[k.strip()] = v.strip().strip('"').strip("'")
    return out

def apply_migration(token: str, ref: str, sql_path: Path) -> tuple[bool, str]:
    sql = sql_path.read_text()
    url = f"https://api.supabase.com/v1/projects/{ref}/database/query"
    body = json.dumps({"query": sql}).encode("utf-8")
    req = urllib.request.Request(
        url,
        data=body,
        headers={
            "Authorization": f"Bearer {token}",
            "Content-Type": "application/json"
        },
        method="POST"
    )
    try:
        with urllib.request.urlopen(req, timeout=60) as resp:
            data = resp.read().decode("utf-8")
            return True, data
    except urllib.error.HTTPError as e:
        return False, f"HTTP {e.code}: {e.read().decode('utf-8')}"
    except Exception as e:
        return False, f"{type(e).__name__}: {e}"

def main() -> int:
    env = load_env(Path(".env.tools.local"))
    token = env.get("SUPABASE_ACCESS_TOKEN")
    ref = env.get("SUPABASE_PROJECT_REF")
    if not token or not ref:
        print("Missing SUPABASE_ACCESS_TOKEN or SUPABASE_PROJECT_REF in .env.tools.local")
        return 1

    migrations = sys.argv[1:] if len(sys.argv) > 1 else [
        "supabase/migrations/20260710120000_reviews.sql",
        "supabase/migrations/20260710130000_canteens.sql",
        "supabase/migrations/20260710140000_uploads_usage.sql",
        "supabase/migrations/20260710150000_merchant_recovery.sql"
    ]

    failures = 0
    for m in migrations:
        p = Path(m)
        print(f"── {p.name}")
        ok, out = apply_migration(token, ref, p)
        status = "✓" if ok else "✗"
        print(f"  {status} {out[:400]}")
        if not ok:
            failures += 1
        print()

    print(f"Done. {len(migrations) - failures}/{len(migrations)} succeeded.")
    return 0 if failures == 0 else 1

if __name__ == "__main__":
    sys.exit(main())

File: scripts/apply-migrations.mjs
// Apply Supabase migrations via the Management API.
// Node 18+ (fetch built in).
import { readFileSync, existsSync } from "node:fs";
import path from "node:path";

function loadEnv(file) {
  const out = {};
  if (!existsSync(file)) return out;
  const text = readFileSync(file, "utf-8");
  for (const raw of text.split(/\r?\n/)) {
    const line = raw.trim();
    if (!line || line.startsWith("#") || !line.includes("=")) continue;
    const [k, ...rest] = line.split("=");
    out[k.trim()] = rest.join("=").trim().replace(/^["']|["']$/g, "");
  }
  return out;
}

async function applyMigration(token, ref, sqlPath) {
  const sql = readFileSync(sqlPath, "utf-8");
  const res = await fetch(`https://api.supabase.com/v1/projects/${ref}/database/query`, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${token}`,
      "Content-Type": "application/json"
    },
    body: JSON.stringify({ query: sql })
  });
  const body = await res.text();
  return { ok: res.ok, status: res.status, body };
}

async function main() {
  const env = loadEnv(".env.tools.local");
  const token = env.SUPABASE_ACCESS_TOKEN;
  const ref = env.SUPABASE_PROJECT_REF;
  if (!token || !ref) {
    console.error("Missing SUPABASE_ACCESS_TOKEN or SUPABASE_PROJECT_REF in .env.tools.local");
    process.exit(1);
  }

  const migrations = process.argv.slice(2).length > 0
    ? process.argv.slice(2)
    : [
        "supabase/migrations/20260710120000_reviews.sql",
        "supabase/migrations/20260710130000_canteens.sql",
        "supabase/migrations/20260710140000_uploads_usage.sql",
        "supabase/migrations/20260710150000_merchant_recovery.sql"
      ];

  let failures = 0;
  for (const m of migrations) {
    const name = path.basename(m);
    process.stdout.write(`── ${name}\n`);
    const { ok, status, body } = await applyMigration(token, ref, m);
    const mark = ok ? "✓" : "✗";
    process.stdout.write(`  ${mark} status=${status} ${body.slice(0, 400)}\n\n`);
    if (!ok) failures += 1;
  }

  process.stdout.write(`Done. ${migrations.length - failures}/${migrations.length} succeeded.\n`);
  process.exit(failures === 0 ? 0 : 1);
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
